core/src/commonMain/kotlin/com/powersync/attachments/AttachmentQueue.kt

```diff
@@ -0,0 +1,489 @@
+     * reappears after being removed. This parameter defines how many archived records are retained.
+     * Records are deleted once the number of items exceeds this value.
+     */
+    private val archivedCacheLimit: Long = 100,
+    /**
+     * Throttles remote sync operations triggering
+     */
+    private val syncThrottleDuration: Duration = 1.seconds,
+    /**
+     * Creates a list of subdirectories in the {attachmentDirectoryName} directory
+     */
+    private val subdirectories: List<String>? = null,
+    /**
+     * Should attachments be downloaded
+     */
+    private val downloadAttachments: Boolean = true,
+    /**
+     * Logging interface used for all log operations
+     */
+    public val logger: Logger = Logger,
+) {
+    public companion object {
+        public const val DEFAULT_TABLE_NAME: String = "attachments"
+        public const val DEFAULT_ATTACHMENTS_DIRECTORY_NAME: String = "attachments"
+    }
+
+    /**
+     * Service which provides access to attachment records.
+     * Use this to:
+     *  - Query all current attachment records
+     *  - Create new attachment records for upload/download
+     */
+    public val attachmentsService: AttachmentService =
+        AttachmentServiceImpl(
+            db,
+            attachmentsQueueTableName,
+            logger,
+            maxArchivedCount = archivedCacheLimit,
+        )
+
+    /**
+     * Syncing service for this attachment queue.
+     * This processes attachment records and performs relevant upload, download and delete
+     * operations.
+     */
+    private val syncingService: SyncingService =
+        SyncingService(
+            remoteStorage,
+            localStorage,
+            attachmentsService,
+            ::getLocalUri,
+            errorHandler,
+            logger,
+            syncThrottleDuration,
+        )
+
+    private var syncStatusJob: Job? = null
+    private val mutex = Mutex()
+
+    public var closed: Boolean = false
+
+    /**
+     * Initialize the attachment queue by
+     * 1. Creating attachments directory
+     * 2. Adding watches for uploads, downloads, and deletes
+     * 3. Adding trigger to run uploads, downloads, and deletes when device is online after being offline
+     */
+    @Throws(PowerSyncException::class, CancellationException::class)
+    public suspend fun startSync(): Unit =
+        runWrappedSuspending {
+            mutex.withLock {
+                if (closed) {
+                    throw Exception("Attachment queue has been closed")
+                }
+                // Ensure the directory where attachments are downloaded, exists
+                localStorage.makeDir(attachmentsDirectory)
+
+                subdirectories?.forEach { subdirectory ->
+                    localStorage.makeDir(Path(attachmentsDirectory, subdirectory).toString())
+                }
+
+                val scope = CoroutineScope(Dispatchers.IO)
```

Comment:
This effectively launches the `syncStatusJob` inside the `GlobalScope` without structured concurrency. I think this should take a `CoroutineScope` from which it inherits. We could:

1. Make `startSync()` an extension function on `CoroutineScope` (`public suspend fun CoroutineScope.startSync()`), a pattern I've repeatedly seen in Kotlin.
2. Make users pass a `CoroutineScope` to the`AttachmentQueue` (it can be the `GlobalScope` by default, but it should at least be possible to override this).
3. Make the coroutine scope for the database implementation `internal` and re-use that here, since we have access to the database.

Then regardless of where we get the outer scope from, we can use `val scope = CoroutineScope(outerScope.newCoroutineContext(Dispatchers.IO))` here.